internal/controlplane/controller.go

```
@@ -0,0 +1,100 @@
+	utilruntime.Must(gatewayv1beta1.AddToScheme(testScheme))
+}
+
+const (
+	inGateControllerName = "k8s.io/ingate"
+)
+
+func Start() error {
+
+	// Register standard Kubernetes types (Pods, Deployments, etc)
+	_ = clientgoscheme.AddToScheme(testScheme)
+
+	// Register Gateway API types (GatewayClass, Gateway, etc.)
+	_ = gatewayv1.AddToScheme(testScheme)
+
+	// Create the ctrl runtime manager
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                 testScheme, // All registered types
+		HealthProbeBindAddress: ":9000",    //needs a flag
+		LeaderElection:         false,      //needs a flag
+		LeaderElectionID:       inGateControllerName,
+		Metrics: metricsserver.Options{
+			BindAddress: ":8080", //needs a flag
+		},
+	})
+	if err != nil {
+		klog.ErrorS(err, "failed to construct InGate manager")
+		return fmt.Errorf("failed to construct InGate manager: %w", err)
+	}
+
+	ctrl.SetLogger(klog.Logger{})
+	log.SetLogger(klog.Logger{})
+	// Add health and readiness probes
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+		klog.ErrorS(err, "unable to set up health check")
+		return err
+	}
+	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+		klog.ErrorS(err, "unable to set up ready check")
+		return err
+	}
+
+	klog.Info("adding gateway class controller")
+	// Create and Add Gateway Class reconciler to manager
+	newGateWayClassReconciler := NewGatewayClassReconciler(mgr)
+
+	err = newGateWayClassReconciler.SetupWithManager(mgr)
+	if err != nil {
+		return err
+	}
+
+	klog.Info("adding gateway controller")
+	// Create and Add Gateway reconciler to manager
+	newGateWayReconciler := NewGatewayReconciler(mgr)
+	err = newGateWayReconciler.SetupWithManager(mgr)
+	if err != nil {
+		return err
+	}
+
+	klog.Info("Starting InGate Manager")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+		klog.Errorf("problem running manager %s", err.Error())
+		os.Exit(1)
```

this os.Exit(1) probably is not required here, instead you can just return the error and call the os.Exit(1) from main function